AdaptiveCardHost - Added "context" property

fabiofranzini opened the pull request:
| Q               | A
| --------------- | ---
| New feature?    | [X]
| Related issues?  | #1145 #1154

#### What's in this Pull Request?

Added SPFx "context" property to "AdaptiveCardHost" control to inject "@context" field in the data object for access context and theme information inside the Adaptive Card using the templating model.

AJIXuMuK commented on src/controls/adaptiveCardHost/AdaptiveCardHostHelpers.ts:
```diff
@@ -1024,4 +1025,27 @@ const setCSSVariables = (domElement: HTMLElement, prefix: string, obj: any) => {
+    let evaluationContext: IEvaluationContext;
+    let context = {
+        theme: theme,
+        aadInfo: (spfxContext?.pageContext?.aadInfo) ? JSON.parse(JSON.stringify(spfxContext?.pageContext?.aadInfo)) : undefined,
```
`JSON.stringify` on classes may lead to issues in future if there some new properties-instances with circular dependencies.
Not ideal way, but not buggy, is to directly specify what properties from the class you want to copy to your context object.

fabiofranzini replied:
Hi @AJIXuMuK , I understand your request of changes and you are right. Let me understand a different way to do that, and I'll push the updates.

fabiofranzini replied:
Hi @AJIXuMuK , I followed your suggestion and fixed how to retrieve and use context properties. This way there are no problems with serialization. Thanks!

fabiofranzini commented:
Hi @AJIXuMuK , I removed the "spPartialLegacyContext" from my previous commit because it's doesn't have sense at this moment. The legacy context from SPFx PageContext is legacy and at this moment it's not thought to use it in any scenario.

I also added a new commit with the code to add the "telemetry" for this control. 🙂

fabiofranzini commented:
Hi @AJIXuMuK, can you check the last commit and give me some feedback on this code? I need to approve this PR before creating a new one 🙂

fabiofranzini commented:
Hi everyone, I'm implementing Adaptive Cards designer as a react control. to do this I'm thinking about how to modify and reuse this control. This PR was to introduce "direct" support to SharePoint context properties, directly in the "data" object.
This, however, after much reasoning, introduces a "link" with the context of SPFx, and if one does not want to use it in this way one would be forced to have to pass the SharePoint context anyway.

That said, I was struggling to change the PR again and introduce not a control property, but a helper that allows you to inject the same properties in a way like this:

```
let dataObject = {"a":1};
<AdaptiveCardHost
  ...
  data={injectSPFxContextData(dataObject)}
  ...
/>
```
before approving this PR, can you tell me your opinion? 🙂
/cc @AJIXuMuK @joelfmrodrigues @joaojmendes

AJIXuMuK commented:
@fabiofranzini - sorry for the long delay.
Maybe we could just make the `context` property optional?
You already have the logic to combine `data` and `context`.
If the `context` is optional, developers won't need to call any additional functions.

What do you think?

fabiofranzini commented:
> @fabiofranzini - sorry for the long delay. Maybe we could just make the `context` property optional? You already have the logic to combine `data` and `context`. If the `context` is optional, developers won't need to call any additional functions.
> 
> What do you think?

Hi @AJIXuMuK, no problems 🙂
Yes, this could be a good compromise...
Let me do this change today and I'll notify you for the review 🙂

fabiofranzini commented:
@AJIXuMuK done! 
I merged another fix from #1154 
Sorry for this, but I had already gone ahead with the change and then I found myself with the code already written.
For the next few times I will avoid doing this and will keep the code separate for each issue.
I have marked the "isUniqueControlInPage" property as obsolete and we no longer use it in the code, so those who have already used the control will not have problems compiling. 🙂

AJIXuMuK commented:
@fabiofranzini - `context` is still marked as required in the docs.
Also - could you please explain the usage of `nano-css`?
Why is that needed instead of component's `scss` or something from Fluent UI?

fabiofranzini commented:
@AJIXuMuK fixed the docs, thanks 🙂
I used `nano-css` to generate dynamically the class names for some styles to apply to the Adaptive Card HostConfig.
This is because the Adaptive Card HostConfig has a property called `cssClassNamePrefix`, and to set the rendering style (I need it mostly to solve the style problem in the Sub Actions menu) I need to have "dynamic" class names. Otherwise, if they were static, I would have the conflict problem if multiple instances of this control are on the same page (this is the way suggested by the Adaptive Cards product group to remove the `isUniqueControlInPage` property)

AJIXuMuK commented:
merged!

fabiofranzini commented:
> merged!

Thanks @AJIXuMuK!
Next step designer 😀
